### ISSUE TITLE
Add a GRPC interface to the notary server

### ISSUE THREAD

**Issue by endophage** [notaryproject/notary#14]
Future feature.

It would be awesome to have a GRPC interface.

**Comment by progrium** [comment]
+1

**Comment by HuKeping** [comment]
Hi @endophage , I'm not very familiar with Notary right now, but when I reading the code of Notary, I found that Notary use both RPC and REST, but why. Isn't the RPC  coupling? 

Clients are required to know procedure names, procedure parameters order, types and count matters, etc. It's not that easy to change procedure signatures without breaking client implementations.

So why not all by REST?

**Comment by endophage** [comment]
@HuKeping the strictness is what makes it desirable. It makes it obvious when a client will be broken. One of the greatest pain points with REST based APIs is that over the course of a year or two, many small changes can break clients without anybody realizing until bug reports start rolling in.

GRPC actually helps uncouple clients and servers because it explicitly and strictly defines their communication interface. This allows any client to interact with any server as long as they each implement one side of the GRPC protocol. Furthermore, as a server dev, I don't have to know anything about the client, I only need to know the communication interface, and vice versa for the client (this can significantly accelerate development efforts). Additionally, GRPC provides tooling to easily compile both client and server libraries for the API in a wide range of popular languages, which helps open up the platform for developers to write compatible clients and servers in their favourite language, that will be able to interact with already existing clients and servers.

I would hazard a guess that all clients written against REST APIs have a few hacks to deal with unintentional weirdness in the API that isn't documented and made it to production accidentally. Each one of those hacks is a very tight coupling to the specific server the client was built to interact with.

At the end of the day, in my mind, it comes down to the question of why should we be less rigorous about code when it's communicating over a network vs via direct function calls.

**Comment by HuKeping** [comment]
Thanks @endophage , I'll take a deeper look into the RPC.

But Notary will maintain two different APIs, although it not a big deal.

Nice to see you  @thaJeztah here, you are really the man who is everywhere ;)

**Comment by thaJeztah** [comment]
@HuKeping haha, you can't escape me!

**Comment by endophage** [comment]
@HuKeping you're right, we'll definitely have to maintain 2 APIs for some period of time, but once the GRPC version is available, we'd deprecate the HTTP version and eventually (a year or more after end to end GRPC release), probably remove it entirely.

**Comment by diogomonica** [comment]
@HuKeping it is also worthy of note the fact that notary-server already uses GRPC as the communication method with notary-signer.

**Comment by jayjhan8** [comment]
Is there an update to this?

**Comment by endophage** [comment]
Hey @jayjhan8, this is a tracking issue that doesn't really have a timeline. We just didn't want to lose. If somebody wanted to submit a PR we'd be happy to review it and accept.
